lib/schedule.js

```diff
@@ -490,15 +490,20 @@ function recurMatch(val, matcher) {
 
 /* Date-based scheduler */
 function runOnDate(date, job) {
-  var now = (new Date()).getTime();
+  var now = Date.now();
   var then = date.getTime();
 
   if (then < now) {
     setImmediate(job);
     return null;
   }
 
-  return lt.setTimeout(job, (then - now));
+  return lt.setTimeout(function() {
+    if (then > Date.now())
```

Comment:
Will Date.now() be the timestamp that the function is run on (as is delayed) or will it be the timestamp the timeout is started?

I think this may cause problems:
* if Date.now() is timestamp of function run: job() is called as expected as then should be <= Date.now()
* if Date.now() is timestamp of starting timeout: then runOnDate is called again recursively

Reply:
`Date.now()` is always the current timestamp when `runOnDate()` is run. Can you elaborate why this can be a problem?

Reply:
So if I understand correctly:
- When `runOnDate()` is run with a date in the future, you setTimeout for `(then - now)` to run the anonymous function
- When that anonymous function executes it compares `then` to the `Date.now()` when `runOnDate()` was originally run
- `then` will be greater than `Date.now()` apart from in the edge case where it is equal to
- `runOnDate()` will be called again and depending on timing will be called immediately or a new timeout started
- If the new timeout is started then that could potentially make another recursive call to `runOnDate()` etc.

Reply:
OK that was my original question, in that case the flow would be:
- When `runOnDate()` is run with a date in the future, you setTimeout for (then - now) to run the anonymous function
- When that anonymous function executes `then` can never greater than `Date.now()`
  - `Date.now()` = `time setTimeoutStarted` + `then` - `time runOnDate() initiallty executed`
  - `time setTimeoutStarted` will be the same or a couple of millis after time `runOnDate()` executed, therefore the Date.now() in the loop will never be less than `then`

If this is correct then the new code doesn't do anything new or is the idea that setTimeout can actually execute sooner than expected?

Reply:
> If this is correct then the new code doesn't do anything new or is the idea that setTimeout can actually execute sooner than expected?

Yes that's the issue. It can happen it executes a few milliseconds before expected, that causes that a job can be triggered twice (a few milliseconds before and then in the expected time). See: https://github.com/node-schedule/node-schedule/issues/263#issuecomment-339431829 for a reproducible test case.

Reply:
Ha OK, thanks for explaining :) I wonder why this happens - clock time changing slightly during execution?